crafting.py

```diff
@@ -0,0 +1,316 @@
+
+    def __eq__(self, other) -> bool:
+        return self.items == other.items
+
+    def __hash__(self) -> int:
+        hsh = 5381
+        for s in self.items:
+            hsh = ((hsh << 5) + hsh) + s
+        return hsh
+
+    def __str__(self) -> str:
+        return self.to_dict().__str__()
+
+    def to_dict(self) -> Dict[str, int]:
+        return {items_by_index[idx]: self.items[idx]
+                for idx in range(len(self.items))}
+
+    @classmethod
+    def from_dict(cls, item_dict: Dict[str, int]) -> 'State':
+        return cls([
+            item_dict.get(item,0) for item in items_by_index
+        ])
+
+class Recipe(NamedTuple):    
+    produces: State
+    consumes: State
+    requires: State
+    cost: int
+
+recipes: Dict[str, Recipe] = {}
+for name, rule in Crafting['Recipes'].items():
+    recipes[name] = Recipe(
+        State.from_dict(rule.get('Produces', {})),
+        State.from_dict(rule.get('Consumes', {})),
+        State.from_dict({item: 1 if req else 0
+                         for item, req in rule.get('Requires', {}).items()}),
+        rule['Time']
+    )   
+
+class Proposition(NamedTuple):
+    item: int
+    at_least: int
+
+def state_propositions(state: State) -> Set[Proposition]:
+    propositions: Set[Proposition] = set()
+    
+    for i in range(len (state.items)) :
+        for k in range(1,state.items[i]+1):
+            propositions.add(Proposition(i,k))
+     
+    return propositions
+
+
+def recipe_to_propositions(recipe: Recipe) -> Set[Proposition]:
+    propositions: Set[Proposition] = set()
+    # G. Do something with recipe.consumes, recipe.produces, and recipe.requires.
+    # Emit, for this recipe, all the propositions entailed by the preconditions and the _minimal_ set of propositions embodied in the preconditions (i.e., don't need to output wood >= 2, wood >= 1, wood >= 0 if the recipe creates 2 wood.)
+    r = state_propositions(recipe.requires)
+    p = state_propositions(recipe.produces)
+    c = state_propositions(recipe.consumes)
+                    
+    propositions |= r     
+    propositions |= c    
+    propositions |= p   
+   
+        
+    return propositions 
+
+    
+def preconditions_satisfied(state: State, recipe: Recipe) -> bool:
+  
+    if ( state >= recipe.requires and state >= recipe.consumes) :
+        return True
+    
+    return False
+
+
+
+def apply_effects(state: State, recipe: Recipe) -> State:
+    # E. How do you make a new state out of a state and a recipe?
+    # Note, DO NOT change state in-place!
+     newstate  =  state - recipe.consumes + recipe.produces
+    
+     return newstate 
+ 
+def see_state(state:State, combinations:List[Set[Proposition]], seen_combinations:Set[FrozenSet[Proposition]]) -> bool:
+    any_new = False
+    state_props = state_propositions(state)
+    for combo in combinations:
+        if combo in seen_combinations:
+            continue 
+        
+        if(state_props.issuperset(combo)) :
```

Comment:
Should this if-statement be nested inside the previous one? Or else what is the point of the `if combo in combinations: continue` check? I think what I have is checking if the combo is not in seen_combinations and then do the superset check - so maybe just converting the superset to an `elif` might help?

Reply:
Nevermind, I just did some research on the `continue` statement. That is good to know! I do feel though that it might help readability to just have one check the not condition maybe.